safeprime/safeprime.go

```diff
@@ -71,16 +78,28 @@ func Generate(bitsize int, stop chan struct{}) (*big.Int, error) {
 	var (
 		one        = big.NewInt(1)
 		two        = big.NewInt(2)
-		max        = new(big.Int).Lsh(one, uint(bitsize)) // 2^bitsize, len bitsize+1
 		twoq       = new(big.Int)
 		twoqone    = new(big.Int)
 		twoexptwoq = new(big.Int)
-		q          *big.Int
-		bitlen     int
+		q          = new(big.Int)
+		bigMod     = new(big.Int)
 		err        error
 		i          int
 	)
 
+	// The algorithm generates a prime q and then returns 2q+1. The latter must have length bitsize,
+	// so the former must be one bit shorter.
+	bitsize--
+
+	// b is the number of bits in the most significant byte of the output that will be part of our q
+	b := uint(bitsize % 8)
```

Comment:
Apparently, gabi's `RandomPrimeInRange` in randomprime.go is already based on crypto/rand. There's quite a bit of code duplication now. Have you checked if we could re-use a bit more of that code?